exp/controllers/awsmachinepool_controller_test.go

```diff
@@ -0,0 +1,469 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+	"k8s.io/utils/pointer"
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+	expinfrav1 "sigs.k8s.io/cluster-api-provider-aws/exp/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/mock_services"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	capierrors "sigs.k8s.io/cluster-api/errors"
+	expclusterv1 "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+var _ = Describe("AWSMachinePoolReconciler", func() {
+	var (
+		reconciler AWSMachinePoolReconciler
+		cs         *scope.ClusterScope
+		ms         *scope.MachinePoolScope
+		mockCtrl   *gomock.Controller
+		ec2Svc     *mock_services.MockEC2MachineInterface
+		asgSvc     *mock_services.MockASGInterface
+		recorder   *record.FakeRecorder
+	)
+
+	BeforeEach(func() {
+		var err error //TODO: check out LogToOutput
+
+		if err := flag.Set("logtostderr", "false"); err != nil {
+			_ = fmt.Errorf("Error setting logtostderr flag")
+		}
+		if err := flag.Set("v", "2"); err != nil {
+			_ = fmt.Errorf("Error setting v flag")
+		}
+
+		klog.SetOutput(GinkgoWriter)
+
+		awsMachinePool := &expinfrav1.AWSMachinePool{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "test",
+			},
+			Spec: expinfrav1.AWSMachinePoolSpec{},
+		}
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "bootstrap-data",
+			},
+			Data: map[string][]byte{
+				"value": []byte("shell-script"),
+			},
+		}
+
+		ms, err = scope.NewMachinePoolScope(
+			scope.MachinePoolScopeParams{
+				Client: fake.NewFakeClient([]runtime.Object{awsMachinePool, secret}...),
```

Comment:
for an integration test, this shouldn't be a NewFakeClient, but instead should use the k8sClient var created in suite_test.go